File: 4196-maximum-pulse-value-after-one-subarray-rotation/maximum-pulse-value-after-one-subarray-rotation.cpp
long long dp[100003][3][2][3];

long long fun(int i,vector<int>&nums,int f,int sg,int pr){
    if(i>=nums.size()){
        if(f==1) return -1e17;
        return 0;
    }

    if(dp[i][f][sg==1?1:0][pr+1]!=-1e18)
        return dp[i][f][sg==1?1:0][pr+1];

    long long crr=1LL*nums[i]*sg;

    long long m=-1e17;

    if(f==0){
        long long a=(-1*crr)+fun(i+1,nums,1,sg,i%2);
        long long b=(crr)+fun(i+1,nums,f,-1*sg,pr);

        m=max(m,a);
        m=max(m,b);
    }
    else if(f==1){
        long long a=(crr)+fun(i+1,nums,f,-1*sg,pr);
        m=max(m,a);

        if(i%2!=pr){
            long long b=crr+fun(i+1,nums,2,sg,-1);
            m=max(m,b);
        }
    }
    else if(f==2){
        long long a=(crr)+fun(i+1,nums,f,-1*sg,pr);
        m=max(m,a);
    }

    return dp[i][f][sg==1?1:0][pr+1]=m;
}


class Solution {
public:
    long long maxValue(vector<int>& nums) {
        for(int i=0;i<100003;i++){
            for(int j=0;j<3;j++){
                for(int k=0;k<2;k++){
                    for(int l=0;l<3;l++){
                        dp[i][j][k][l]=-1e18;
                    }
                }
            }
        }

        return fun(0,nums,0,1,-1);
    }
};

File: 4196-maximum-pulse-value-after-one-subarray-rotation/maximum-pulse-value-after-one-subarray-rotation.java
import java.util.Arrays;

class Solution {
    private long[][][][] dp;

    private long fun(int i, int[] nums, int f, int sg, int pr) {
        if (i >= nums.length) {
            if (f == 1) return (long) -1e17;
            return 0;
        }

        int sgIdx;
        if (sg == 1) {
            sgIdx = 1;
        } else {
            sgIdx = 0;
        }

        if (dp[i][f][sgIdx][pr + 1] != (long) -1e18) {
            return dp[i][f][sgIdx][pr + 1];
        }

        long crr = 1L * nums[i] * sg;
        long m = (long) -1e17;

        if (f == 0) {
            long a = (-1 * crr) + fun(i + 1, nums, 1, sg, i % 2);
            long b = (crr) + fun(i + 1, nums, f, -1 * sg, pr);

            m = Math.max(m, a);
            m = Math.max(m, b);
        } 
        else if (f == 1) {
            long a = (crr) + fun(i + 1, nums, f, -1 * sg, pr);
            m = Math.max(m, a);

            if (i % 2 != pr) {
                long b = crr + fun(i + 1, nums, 2, sg, -1);
                m = Math.max(m, b);
            }
        } 
        else if (f == 2) {
            long a = (crr) + fun(i + 1, nums, f, -1 * sg, pr);
            m = Math.max(m, a);
        }

        return dp[i][f][sgIdx][pr + 1] = m;
    }

    public long maxValue(int[] nums) {
        int n = nums.length;
        dp = new long[n + 1][3][2][3];

        for (int i = 0; i <= n; i++) {
            for (int j = 0; j < 3; j++) {
                for (int k = 0; k < 2; k++) {
                    Arrays.fill(dp[i][j][k], (long) -1e18);
                }
            }
        }

        return fun(0, nums, 0, 1, -1);
    }
}